pyquil/operator_estimation.py

```diff
@@ -24,6 +26,105 @@
 
 
 @dataclass(frozen=True)
+class _OneQState:
+    """
+    A description of a named one-qubit quantum state.
+
+    This can be used to generate pre-rotations for quantum process tomography. For example,
+    X0_14 will generate the +1 eigenstate of the X operator on qubit 14. X1_14 will generate the
+    -1 eigenstate. SIC0_14 will generate the 0th SIC-basis state on qubit 14.
+    """
+    label: str
+    index: int
+    qubit: int
+
+    def __str__(self):
+        return f'{self.label}{self.index}_{self.qubit}'
+
+    @classmethod
+    def from_str(cls, s):
+        ma = re.match(r'\s*(\w+)(\d+)_(\d+)\s*', s)
+        if ma is None:
+            raise ValueError(f"Couldn't parse '{s}'")
+        return _OneQState(
+            label=ma.group(1),
+            index=int(ma.group(2)),
+            qubit=int(ma.group(3)),
+        )
+
+
+@dataclass(frozen=True)
+class TensorProductState:
+    """
+    A description of a multi-qubit quantum state that is a tensor product of many _OneQStates
+    states.
+    """
+    states: Tuple[_OneQState]
+
+    def __init__(self, states):
+        object.__setattr__(self, 'states', tuple(states))
+
+    def __mul__(self, other):
+        return TensorProductState(self.states + other.states)
+
+    def __str__(self):
+        return ' * '.join(str(s) for s in self.states)
```

Comment:
This looks a little funny if a vacuum state is passed in with anything else. Maybe it would be understood that this doesn't make sense in the first place.

Reply:
yeah, the general idea is that you ought never have bare `OneQState`s. It's always wrapped in a (length-1) TensorProductState